runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/artifact/ArtifactStagingService.java

```diff
@@ -0,0 +1,606 @@
+       *
+       * @param index a monotonically increasing index, which provides uniqueness
+       * @param environment the environment id
+       * @param artifact the artifact itself
+       */
+      private String createFilename(
+          int index, String environment, RunnerApi.ArtifactInformation artifact) {
+        String path;
+        try {
+          if (artifact.getRoleUrn().equals(ArtifactRetrievalService.STAGING_TO_ARTIFACT_URN)) {
+            path =
+                RunnerApi.ArtifactStagingToRolePayload.parseFrom(artifact.getRolePayload())
+                    .getStagedName();
+          } else if (artifact.getTypeUrn().equals(ArtifactRetrievalService.FILE_ARTIFACT_URN)) {
+            path = RunnerApi.ArtifactFilePayload.parseFrom(artifact.getTypePayload()).getPath();
+          } else if (artifact.getTypeUrn().equals(ArtifactRetrievalService.URL_ARTIFACT_URN)) {
+            path = RunnerApi.ArtifactUrlPayload.parseFrom(artifact.getTypePayload()).getUrl();
+          } else {
+            path = "artifact";
+          }
+        } catch (InvalidProtocolBufferException exn) {
+          throw new RuntimeException(exn);
+        }
+        // Limit to the last contiguous alpha-numeric sequence. In particular, this will exclude
+        // all path separators.
+        List<String> components = Splitter.onPattern("[^A-Za-z-_.]]").splitToList(path);
+        String base = components.get(components.size() - 1);
+        return clip(String.format("%d-%s-%s", index, clip(environment, 25), base), 100);
+      }
+
+      private String clip(String s, int maxLength) {
+        return s.length() < maxLength ? s : s.substring(0, maxLength);
+      }
+
+      @Override
+      public void onError(Throwable throwable) {
+        stagingExecutor.shutdownNow();
+        LOG.error("Error staging artifacts", throwable);
+        state = State.ERROR;
+      }
+
+      @Override
+      public void onCompleted() {
+        assert state == State.DONE;
+      }
+    };
+  }
+
+  @Override
+  public void close() throws Exception {
+    // Nothing to close.
+  }
+
+  /**
+   * Lazily stages artifacts by letting an ArtifactStagingService resolve and request artifacts.
+   *
+   * @param retrievalService an ArtifactRetrievalService used to resolve and retrieve artifacts
+   * @param stagingService an ArtifactStagingService stub which will request artifacts
+   * @param stagingToken the staging token of the job whose artifacts will be retrieved
+   * @throws InterruptedException
+   * @throws IOException
+   */
+  public static void offer(
+      ArtifactRetrievalService retrievalService,
+      ArtifactStagingServiceGrpc.ArtifactStagingServiceStub stagingService,
+      String stagingToken)
+      throws InterruptedException, IOException {
+    StagingRequestObserver requestObserver = new StagingRequestObserver(retrievalService);
+    requestObserver.responseObserver =
+        stagingService.reverseArtifactRetrievalService(requestObserver);
```

Comment:
I had problems like this as well when using gRPC API. You should be able to pass in stagingService to the StagingRequestObserver constructor allowing the responseObserver to be final since you should be able to pass `this` to reverseArtifactRetrievalService